File: trip-kit-booking/src/main/java/com/skedgo/tripkit/booking/viewmodel/OptionFieldViewModel.java
package com.skedgo.tripkit.booking.viewmodel;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.text.TextUtils;

import com.skedgo.tripkit.common.rx.Var;
import com.skedgo.tripkit.booking.OptionFormField;

import io.reactivex.Flowable;
import org.apache.commons.collections4.CollectionUtils;

import java.util.List;

public class OptionFieldViewModel implements TwoLineViewModel {
  private final OptionFormField optionField;
  private Var<OptionFieldViewModel> onValueSelected = Var.create();
  private int selectedIndex;

  protected OptionFieldViewModel(@NonNull OptionFormField optionField) {
    this.optionField = optionField;
    this.onValueSelected.put(this);
    this.selectedIndex = -1;
  }

  public static OptionFieldViewModel create(@NonNull OptionFormField optionField) {
    return new OptionFieldViewModel(optionField);
  }

  public Flowable<OptionFieldViewModel> onValueSelected() {
    return onValueSelected.observe();
  }

  public List<OptionFormField.OptionValue> getAllValues() {
    return optionField.getAllValues();
  }

  public int getSelectedIndex() {
    if (selectedIndex == -1) {
      selectedIndex = findSelectedIndex();
    }
    return selectedIndex;
  }

  public void select(int valueIndex) {
    selectedIndex = valueIndex;
    final List<OptionFormField.OptionValue> allValues = optionField.getAllValues();
    if (CollectionUtils.isEmpty(allValues)
        && (valueIndex < 0 || valueIndex >= allValues.size())) {
      return;
    }

    final OptionFormField.OptionValue newSelectedValue = allValues.get(valueIndex);
    optionField.setValue(newSelectedValue);
    onValueSelected.put(this);
  }

  @Override
  public String getPrimaryText() {
    final OptionFormField.OptionValue selectedValue = optionField.getValue();
    return selectedValue != null ? selectedValue.getTitle() : null;
  }

  @Nullable
  @Override
  public String getSecondaryText() {
    final OptionFormField.OptionValue selectedValue = optionField.getValue();
    return selectedValue != null ? selectedValue.getValue() : null;
  }

  private int findSelectedIndex() {
    final OptionFormField.OptionValue selectedValue = optionField.getValue();
    final List<OptionFormField.OptionValue> allValues = optionField.getAllValues();
    if (selectedValue != null && CollectionUtils.isNotEmpty(allValues)) {
      for (int i = 0; i < allValues.size(); i++) {
        if (TextUtils.equals(allValues.get(i).getValue(), selectedValue.getValue())) {
          return i;
        }
      }
    }
    return 0;
  }
}

File: trip-kit-booking/src/main/java/com/skedgo/tripkit/booking/viewmodel/OptionFieldViewModel.kt
package com.skedgo.tripkit.booking.viewmodel

import com.skedgo.tripkit.booking.OptionFormField
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.subjects.BehaviorSubject

class OptionFieldViewModel(private val optionField: OptionFormField) : TwoLineViewModel {

    private val onValueSelected: BehaviorSubject<OptionFieldViewModel> = BehaviorSubject.create()
    private var selectedIndex: Int = -1

    init {
        onValueSelected.onNext(this)
    }

    companion object {
        fun create(optionField: OptionFormField): OptionFieldViewModel {
            return OptionFieldViewModel(optionField)
        }
    }

    fun onValueSelected(): Flowable<OptionFieldViewModel> {
        return onValueSelected.hide().toFlowable(BackpressureStrategy.BUFFER)
    }

    fun getAllValues(): List<OptionFormField.OptionValue> {
        return optionField.allValues
    }

    fun getSelectedIndex(): Int {
        if (selectedIndex == -1) {
            selectedIndex = findSelectedIndex()
        }
        return selectedIndex
    }

    fun select(valueIndex: Int) {
        val allValues = optionField.allValues
        if (allValues.isNullOrEmpty() || valueIndex < 0 || valueIndex >= allValues.size) {
            return
        }

        selectedIndex = valueIndex
        val newSelectedValue = allValues[valueIndex]
        optionField.setValue(newSelectedValue)
        onValueSelected.onNext(this)
    }

    override fun getPrimaryText(): String? {
        val selectedValue = optionField.getValue()
        return selectedValue?.title
    }

    override fun getSecondaryText(): String? {
        val selectedValue = optionField.getValue()
        return selectedValue?.value
    }

    private fun findSelectedIndex(): Int {
        val selectedValue = optionField.getValue()
        val allValues = optionField.allValues
        if (selectedValue != null && allValues.isNotEmpty()) {
            for (i in allValues.indices) {
                if (allValues[i].value == selectedValue.value) {
                    return i
                }
            }
        }
        return 0
    }
}